plugins/woocommerce/includes/class-wc-structured-data.php

```diff
@@ -257,9 +257,25 @@ public function generate_product_data( $product = null ) {
 				);
 			}
 
+			$stock_status = $product->get_stock_status();
+
+			switch ( $stock_status ) {
+				case 'instock':
+					$stock_status_schema = 'InStock';
+					break;
+				case 'outofstock':
+                    $stock_status_schema = 'OutOfStock';
+					break;
+				case "onbackorder":
+                    $stock_status_schema = 'BackOrder';
+					break;
+				default:
+                    $stock_status_schema = '';
```

Comment:
I don't think we should allow an empty `$stock_status_schema` as that might result in broken JSON. Also, looking at how this is working right now, I can see that we are really using the result from [`is_in_stock()`](https://github.com/woocommerce/woocommerce/blob/b0f042e252cd7adeabfe6c131734c3b0aa40bd36/plugins/woocommerce/includes/abstracts/abstract-wc-product.php#L1670-L1672) which involves a filter.

We don't want to lose that flexibility (for other plugins to use that filter to change the "in stock" status) nor affect backwards compatibility, so I'd suggest we do things slightly different.

How about instead of a `case` we do a check for `is_in_stock()` and set `$stock_status_schema` to either `InStock` or `BackOrder` if the stock status is explicitly "backorder". Otherwise, we set `$stock_status_schema` to `OutOfStock`.

That way we preserve current behavior and also reduce the chances that we end up with an empty `$stock_status_schema`.

For example, something along these lines might work:

```php
if ( $product->is_in_stock() ) {
	$stock_status_schema = ( 'onbackorder' === $product->get_stock_status() ) ? 'BackOrder' : 'InStock';
} else {
	$stock_status_schema = 'OutOfStock';
}
```

What are your thoughts?